### PR TITLE
Yolo benchmarking

### PR DESCRIPTION
**Pull request by bfineran**

### CODE REVIEW

**Review by markurtz — DISMISSED** [review]

File: integrations/ultralytics/deepsparse/benchmark.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+
+def _load_file_paths(dataset_path: str) -> List[str]:
+    # from yolov5/utils/datasets.py
+    path = str(Path(dataset_path).absolute())  # os-agnostic absolute path
+    if '*' in path:
+        files = sorted(glob.glob(path, recursive=True))  # glob
+    elif os.path.isdir(path):
+        files = sorted(glob.glob(os.path.join(path, '*.*')))  # dir
+    elif os.path.isfile(v):
```
Comment (bfineran) [inline integrations/ultralytics/deepsparse/benchmark.py:158]:
```suggestion
    elif os.path.isfile(path):
```

File: integrations/ultralytics/deepsparse/benchmark.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        files = [path]  # files
+    else:
+        raise Exception(f"ERROR: {p} does not exist")
```
Comment (bfineran) [inline integrations/ultralytics/deepsparse/benchmark.py:161]:
```suggestion
        raise Exception(f"ERROR: {path} does not exist")
```

**Review by markurtz — DISMISSED** [review]

File: integrations/ultralytics/deepsparse/benchmark.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    )
+
+    results = BenchmarkResults()
+    progress_bar = tqdm(total=args.num_iterations)
+    for iteration, batch in enumerate(data_loader):
+        iter_start = time.time()
+
+        # pre-processing
+        batch = [cv2.imread(filepath) for filepath in batch]
+        inputs = [preprocess_images(batch)]
```
Comment (bfineran) [inline integrations/ultralytics/deepsparse/benchmark.py:208]:
```suggestion
        inputs = [preprocess_images(batch, fp32=not args.quantized_inputs)]
```

**Review by kevinaer — DISMISSED** [review]

**Review by markurtz — DISMISSED** [review]

**Comment by bfineran** [comment]
Planned changes:
* move batching to dataloader
* torch GPU and ORT backend support

File: integrations/ultralytics/deepsparse/client.py [hunk 4]
```diff
@@ -43,21 +43,19 @@ def detect(
...
+            run the detection model on. Number of images should equal model batch size
         :return: list of post-processed object detection results for each image including
             class label, likelihood, and bounding box coordinates
         """
         if not isinstance(images, List):
             images = [images]
-        images = [
-            cv2.imread(image) if isinstance(image, str) else image for image in images
-        ]
+        images = numpy.stack([load_image(image) for image in images])
```
Comment (mgoin) [inline integrations/ultralytics/deepsparse/client.py:53]: does `len` work on numpy arrays?

**Review by mgoin — DISMISSED** [review]
LGTM just one possible nit

File: integrations/ultralytics/deepsparse/client.py (hunk 4 above)
Comment (bfineran) [reply integrations/ultralytics/deepsparse/client.py:53]: yes, where does `len` factor in here?

Comment (mgoin) [reply integrations/ultralytics/deepsparse/client.py:53]: sorry, i made this on the line below and it selected this instead
```
print(f"Sending batch of {len(images)} images for detection to {self._url}")
```

File: integrations/ultralytics/deepsparse/benchmark.py [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+from pathlib import Path
+from typing import Any, Iterable, List, Optional, Tuple, Union
+
+import numpy
+import onnx
+import onnxruntime
+import torch
+from tqdm.auto import tqdm
+
+import cv2
```
Comment (markurtz) [inline integrations/ultralytics/deepsparse/benchmark.py:108]: Need to add cv2 to known third party for setup config and then run make style again

File: integrations/ultralytics/deepsparse/benchmark.py [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+        "--fp16",
+        help=("Set flag to execute torch benchmark in half precision (fp16)"),
+        action="store_true",
+    )
+    parser.add_argument(
+        "--device",
+        type=_parse_device,
+        default="cpu",
+        help=(
+            "Device to benchmark the model with. Default is cpu. Non cpu benchmarking "
```
Comment (markurtz) [inline integrations/ultralytics/deepsparse/benchmark.py:220]: Can we give an example for the torch use case here? Also, looks like this is the device id going off of the parse device function, can we support cpu and cuda for torch and default to first device? They could still control device outside of this through standard commands and I feel most users will spin this up on single gpu machines

**Review by markurtz — COMMENTED** [review]
Left a few comments

File: integrations/ultralytics/deepsparse/client.py (hunk 4 above)
Comment (bfineran) [reply integrations/ultralytics/deepsparse/client.py:53]: ah, thanks.  images is a list at this point so should be fine.

File: integrations/ultralytics/deepsparse/benchmark.py (hunk 6 above)
Comment (bfineran) [reply integrations/ultralytics/deepsparse/benchmark.py:220]: sounds good.  will make the torch default `0` if cuda is available and add some documentation

**Review by mgoin — APPROVED** [review]

**Review by markurtz — APPROVED** [review]

**Pull request merged**